### PR TITLE
Group functionality

### PR DESCRIPTION
**Pull request by myang0**
## Why?

Group functionality must be added, hence this PR has been created.  As the group functionality is not 100% complete yet, more PRs like this will be made.

## What?

Preliminary changes such as the creation of the Group model for testing, and more configuration options for setting up groups.  Creation of the `scim_groups_controller`, as well as tests for it.  The GET method has been implemented, and tests for it have been created.

### CODE REVIEW

File: spec/dummy/app/models/group.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Group < ActiveRecord::Base
+  has_and_belongs_to_many :users
```
Comment (bj-mcduck) [inline spec/dummy/app/models/group.rb]: should use has_many through
```suggestion
  has_many :users, through: :groups_users
  has_many :groups_users
```

File: spec/dummy/app/models/user.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 class User < ApplicationRecord
   belongs_to :company
 
+  has_and_belongs_to_many :groups
```
Comment (bj-mcduck) [inline spec/dummy/app/models/user.rb]: never use `has_and_belongs_to_many`, just imagine that method no longer exists
use the has many through approach in its place

File: spec/factories/groups_users.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :groups_user do
+    user { nil }
+    group { nil }
```
Comment (bj-mcduck) [inline spec/factories/groups_users.rb]:
```suggestion
    user
    group
```

File: app/controllers/concerns/scim_rails/response.rb [hunk 4]
```diff
@@ -24,6 +24,23 @@ def json_scim_response(object:, status: :ok, counts: nil)
       end
     end
 
+    # TODO: Cut down on redundant code in this file
```
Comment (bj-mcduck) [inline app/controllers/concerns/scim_rails/response.rb]: 🌵 try to have comments right beside whatever it is they're commenting on.
This one is kinda floating off by itself.
If it's about the method, avoid the extra empty line.
If it's about the class, put it at the top above the class definition

File: app/controllers/concerns/scim_rails/response.rb [hunk 5]
```diff
@@ -24,6 +24,23 @@ def json_scim_response(object:, status: :ok, counts: nil)
...
+
+    def json_scim_group_response(object:, status: :ok, counts: nil)
+      case params[:action]
+      when "index"
+        render \
```
Comment (bj-mcduck) [inline app/controllers/concerns/scim_rails/response.rb]: 🌵 🌵 the usage of `\` here is odd

File: app/controllers/concerns/scim_rails/response.rb [hunk 6]
```diff
@@ -24,6 +24,23 @@ def json_scim_response(object:, status: :ok, counts: nil)
...
+          json: list_group_response(object, counts),
+          status: status,
+          content_type: CONTENT_TYPE
+      when "show", "create", "put_update", "patch_update"
+        render \
+          json: group_response(object),
+          status: status,
+          content_type: CONTENT_TYPE
+      end
```
Comment (bj-mcduck) [inline app/controllers/concerns/scim_rails/response.rb]: seems like the only difference is the thing getting rendered here.
cleaner to use the case/switch to set a var for the result and have one render block

File: app/controllers/scim_rails/application_controller.rb [hunk 7]
```diff
@@ -35,5 +35,61 @@ def authenticate_with_oauth_bearer
...
+          found_path = path_for(attribute, object, [*path, key])
+          return found_path if found_path
+        end
+        nil
+      when Array
+        at_path.each_with_index do |value, index|
+          found_path = path_for(attribute, object, [*path, index])
+          return found_path if found_path
+        end
+        nil
```
Comment (bj-mcduck) [inline app/controllers/scim_rails/application_controller.rb:67]: 🌵 What are these `nil` fallbacks doing for you?

File: app/controllers/scim_rails/scim_groups_controller.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      json_scim_response(object: group)
+    end
+
+    def put_update
+
+    end
+
+    def patch_update
+
+    end
```
Comment (bj-mcduck) [inline app/controllers/scim_rails/scim_groups_controller.rb]: 🌵 could remove, or put a comment inside one with either `#TODO ...` or [`#NOOP`](https://en.wikipedia.org/wiki/NOP_(code))

File: app/controllers/scim_rails/scim_groups_controller.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    private
+
+    def permitted_group_params
+      params.permit()
```
Comment (bj-mcduck) [inline app/controllers/scim_rails/scim_groups_controller.rb]: intentional? seems quite wrong

File: app/controllers/concerns/scim_rails/response.rb (hunk 5 above)
Comment (myang0) [reply app/controllers/concerns/scim_rails/response.rb]: This was something I moved, it was in the original code

File: app/controllers/scim_rails/application_controller.rb (hunk 7 above)
Comment (myang0) [reply app/controllers/scim_rails/application_controller.rb:67]: This was also something I only moved around

File: spec/controllers/scim_rails/scim_groups_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        expect(response.status).to eq(200)
+      end
+
+      context 'when less than 100 existing groups' do
+        let(:total_group_count) { 10 }
+
+        let(:returned_resource) { response_body["Resources"].first }
+
+        it 'returns all results' do
+          create_list(:group, total_group_count, users: user_list, company: company)
```
Comment (bj-mcduck) [inline spec/controllers/scim_rails/scim_groups_controller_spec.rb]: move to a let in the context, then you could do the same before block for the request

File: spec/controllers/scim_rails/scim_groups_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+          get :index
+          expect(response_body.dig("schemas", 0)).to eq "urn:ietf:params:scim:api:messages:2.0:ListResponse"
+          expect(response_body["totalResults"]).to eq(total_group_count)
+        end
+
+        let(:group_name) { Faker::Games::Pokemon.location }
```
Comment (bj-mcduck) [inline spec/controllers/scim_rails/scim_groups_controller_spec.rb]: group all lets at the top of their context/describe block

File: spec/controllers/scim_rails/scim_groups_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+            get :index, {
+              filter: "displayName eq #{search_term}"
+            }
+  
+            expect(response_body["totalResults"]).to eq(1)
+            expect(response_body["Resources"].count).to eq(1)
+            expect(returned_resource["displayName"]).to eq(search_term)
+          end
+
+          let(:unfound_search_term) { search_term[0, search_term.length - 2] }
```
Comment (bj-mcduck) [inline spec/controllers/scim_rails/scim_groups_controller_spec.rb]: group at top

File: spec/controllers/scim_rails/scim_groups_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials("unauthorized","123456")
+        expect(response.status).to eq 401
+      end
+    end
+
+    context "when authorized" do
+      before :each do
+        http_login(company)
+      end
+    end
```
Comment (bj-mcduck) [inline spec/controllers/scim_rails/scim_groups_controller_spec.rb]: seems like nothing is in here

File: spec/controllers/scim_rails/scim_groups_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
```
Comment (bj-mcduck) [inline spec/controllers/scim_rails/scim_groups_controller_spec.rb]: same

File: spec/factories/groups.rb [hunk 15]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :group do
+    company {}
+    users { [] }
+
+    display_name { "MyString" }
```
Comment (bj-mcduck) [inline spec/factories/groups.rb]: 🌵 use faker, you already added it

File: spec/factories/groups.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :group do
+    company {}
+    users { [] }
```
Comment (bj-mcduck) [inline spec/factories/groups.rb]: I think you can just not assign these like the other factory you have. These both seem like defaults
```suggestion
    company
    users
```

Comment (myang0) [reply spec/factories/groups.rb]: > I think you can just not assign these like the other factory you have. These both seem like defaults

We have decided to leave this as is for now, for some reason when we change it to
```ruby
users
```
we get an error telling us `Trait not registered: "users"` whenever we try to create a `group`.  On the other hand, if we do
```ruby
users { }
```
we get and error saying `undefined method 'each' for nil:NilClass` if we try to create a group where we don't explicitly set the `users` field, like this: `create(:group, display_name: search_term, company: company)`

**Pull request merged**